### ISSUE TITLE
Yarn requires yarn release .cjs file to be committed to git repo

### ISSUE THREAD

**Issue by ildella** [yarnpkg/berry#5748]
### Self-service

- [X] I'd be willing to implement a fix

### Describe the bug

When I upgrade a project to use yarn canary, it downloads the yarn release file in a local project folder, and writes the path in the .yarnrc.yml. The upgrade to yarn v4 is then successful. 

I do not push the released file to git as I consider a very bad practice to pollute all my repositories with dependencies file of any kind.

Then when someone else tries to run yarn, it fails as any yarn command will look for the local file - which does not exists on git. 

It should be downloading it if there's no local copy. 

### To reproduce

1. upgrade a yarn 1.x project to yarn canary following the step by step guide: https://yarnpkg.com/migration/guide
2. gitignore the whole .yarn/* folder
3. now everything works, commit and push
4. someone else - maybe the github workflow - checks out the project and try to run yarn - really any command even --version:

```shell
node:internal/modules/cjs/loader:1051
  throw err;
  ^

Error: Cannot find module '/{path-to-project}/.yarn/releases/yarn-4.0.0-rc.51.cjs'
    at Module._resolveFilename (node:internal/modules/cjs/loader:1048:15)
    at Module._load (node:internal/modules/cjs/loader:901:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:83:12)
    at node:internal/main/run_main_module:23:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

Node.js v20.7.0
```

### Environment

```shell
System:
    OS: Linux 6.2 Ubuntu 23.04 23.04 (Lunar Lobster)
    CPU: (8) x64 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
  Binaries:
    Node: 20.7.0 - /tmp/xfs-a472fb66/node
    Yarn: 4.0.0-rc.51 - /tmp/xfs-a472fb66/yarn
    npm: 10.1.0 - ~/n/bin/npm
  npmPackages:
    jest: ^29.7.0 => 29.7.0
```

**Comment by ildella** [comment]
Possible solutions:

1. yarn checks `yarnPath` file to exist, and instead of failing just download it.
2. there is a yarn command to download the desired release in a systm folder, like $HOME/.yarn/releases... 

Second option would be great, but we must be sure it can work on CI, like github workflows.

**Comment by polarathene** [comment]
Yes the docs need to be updated. I think many others agree with you that committing versions of yarn and the cache (_with good intent of reproducible builds_) is not good for the git repo.

At the very least git LFS would be more appropriate to keep the core repo history light-weight. It's less of an issue if you want to archive the project, but I don't see the appeal otherwise as git is not really meant to manage binaries like this (_hence git lfs, but that's additional setup_).

---

Your solution to use `corepack prepare` will work for a global yarn on your system.

However for other devs they should have enabled corepack (`corepack enable`) and your `package.json` should have a `packageManager` field that references what version of yarn to use `yarn@3.6.3` for example (_`stable` is not permitted here_) and whenever your project is used, corepack will ensure it's run with the appropriate release of yarn for that project.

Related: https://github.com/yarnpkg/berry/discussions/5744#discussioncomment-7065602

**Comment by ildella** [comment]
Clear. 

The problem I see is that the guide are:

1. It explain how to migrate a single project, rather than upgrading yarn for the system. That's it should be made more clear.
2. Following that guide, user will end up either committing the /release file to git or having failures. Both scenario are unlikely to be welcome. 

This is what a guide to install yarn 2+ at system level could be:

```
    Make sure you're using Node 18+
    Uninstall all existing versions of yarn from your system.
    Run `corepack enable`. See [Corepack](https://nodejs.org/api/corepack.html) for reference. 
    Run `corepack prepare yarn@stable --activate`  to install the desired version. (link to available versions)
    
Now each project must (can?) set which version of yarn to use adding a line like this to package.json:

"packageManager": "yarn@4.0.0-rc.50"

or

"packageManager": "yarn@3.6.3".

At this point in an individual project just run `yarn install` and the project will be upgraded. 
A new file `.yarnrc.yml` will be generated at this stage, and should be published to the version control system.
```

That will allow user to keep old 1.x project as they are - I believe - and smoothly transition just the ones they wish.

**Comment by arcanis** [comment]
> It explain how to migrate a single project, rather than upgrading yarn for the system. That's it should be made more clear.

That's because we don't want you to install Yarn at the system level. We want it to be managed per-project, like any other dependency.

**Comment by ildella** [comment]
But yarn is not like any other dependency. And installing it that way force users to have yarn actual bundle in their VCS, which is bad.

Other dependencies works fine in a folder which is never on VCS.

What are the motivations behind this choice?

**Comment by koshic** [comment]
> But yarn is not like any other dependency.

Why? Every project (scripts, .yarnrc, yarn.lock, etc.) depends on the particular yarn version. Keeping yarn in VCS eliminates all 'global installation' risks, and don't force developers to dance around upgrade / downgrade when switching projects. We already have such experience with Node - 100% backward compatibility is impossible on the long run, and we have to use tools like nvm to choose appropriate version.

Even if sometimes yarn way is a bit... strange / uncommon, ability to 'lock' package manager is a very important feature. Sure, corepack can provide more or less similar experience, but it still requires manual activation and has an inconvenient offline mode.

**Comment by arcanis** [comment]
> And installing it that way force users to have yarn actual bundle in their VCS, which is bad.

You say it's bad, I say "I've been doing it for the past seven years and git is perfectly happy with it". Yarn is 2MB, there's absolutely no problem whatsoever keeping it in the VCS - even perfwise, it has no practical impact on its own _(if you check-in the `.yarn/cache` folder as well it's another story, but that's orthogonal and entirely optional)_.

**With that being said,** you're not the first one raising such concerns, so from the 4.0 forward we are moving towards making Corepack the preferred way to install Yarn - but that's still on a per-project basis (ie with the [`packageManager` field](https://nodejs.org/api/packages.html#packagemanager)), for the reasons described by @koshic.

So once we release the 4.0 stable (which should happen soon), `yarn init -2` will set this field, but won't add the Yarn binary inside the newly created project anymore.

**Comment by polarathene** [comment]
> we don't want you to install Yarn at the system level. We want it to be managed per-project, like any other dependency.
> from the 4.0 forward we are moving towards making Corepack the preferred way to install Yarn - but that's still on a per-project basis

I'm fully in support of `packageManager` in `package.json`, but I don't know why you should hide how to use yarn outside of a project (eg for `yarn dlx`).

I think it's worth highlighting `corepack prepare yarn@stable --activate`, while making the `packageManager` approach the endorsed approach for projects.

> I say "I've been doing it for the past seven years and both git and my coworkers are perfectly happy with it".
> Yarn is 2MB, there's absolutely no problem whatsoever keeping it in the VCS - even perfwise, it has no practical impact on its own

- If you commit all 26 releases on Github since Aug 2019, and we average those at 2MB each, that's 52MB, for a 2MB dependency?
- If you also were doing that with the v4 RC releases, that's over 50, another 100MB.

Both aren't that realistic, but shows over time how that can build up depending how often you'd bump that dependency in the project.

You could just as easily keep an archive of those releases out of the project or use git LFS. In your projects it may not be an issue but it's generally wasteful lumping in such data into your git history.

There's plenty of tooling out there that can get the right version and automate that for you too if you need that (_as mentioned with nvm as an example_). Why not also commit Node and everything else for a full reproducible build as these can be important as time passes too 🤷‍♂️ 

> (if you check-in the `.yarn/cache` folder as well it's another story, but that's orthogonal and entirely optional).

This gets a pass because of how much more excessive it is in size? It's the same problem, you just experience it sooner.

**Comment by arcanis** [comment]
> I'm fully in support of `packageManager` in `package.json`, but I don't know why you should hide how to use yarn outside of a project (eg for `yarn dlx`).

Corepack has a "transparent" mechanism - some commands ([currently only `dlx`](https://github.com/nodejs/corepack/blob/main/config.json#L97-L101)) are run against Yarn 2+, even if 1.x is the global binary. It's not perfect, but it can be improved!

> I think it's worth highlighting `corepack prepare yarn@stable --activate`, while making the `packageManager` approach the endorsed approach for projects.

It's fine to add it to the documentation - but we should make it clear that this is discouraged in general because [reasons described; basically a small FAQ answering the points from this thread]. Then people decide on their own (clearly, the current wording isn't clear enough).

> Both aren't that realistic, but shows over time how that can build up depending how often you'd bump that dependency in the project.

_(I doubt we'll convince each other, so I don't try to address the "is it good or bad" aspect, just the practicality of it)_

Exactly, they aren't realistic - in practice I wouldn't expect you to upgrade Yarn that often. I checked on our work repository, in the past 4y we upgraded it 41 times - and I'm a Yarn maintainer, so sometimes I update in-between releases!

Using LFS for that would have been adding a *significant* hurdle (everyone using the package needs a plugin, for starter) to solve what's essentially a non-problem.

> There's plenty of tooling out there that can get the right version and automate that for you too if you need that (as mentioned with nvm as an example)

And Corepack! Which I actually designed, advocated for, and implemented. I agree that not having to check-in the binary is the best situation - but it wasn't possible until recently, and in the meantime checking-in the binary was actually reasonable.

**Comment by ildella** [comment]
> So once we release the 4.0 stable (which should happen soon), `yarn init -2` will set this field, but won't add the Yarn binary inside the newly created project anymore.

So let me recap: 

1. there will be a fix before final v4 is released so that the per-project approach - with no `corepack prepare` - will NOT require any binary in the VCS
2. The package.json entry will guarantee that regardless one has yarn installed globally or not, the proper version will run a specific project. 
3. the reccomended way will be per-project. With point #1 is actually ok as it won't have potentially undesirable side effects later. 

I think it's a win :)

**Comment by arcanis** [comment]
> there will be a fix before final v4 is released so that the per-project approach - with no corepack prepare - will NOT require any binary in the VCS

It already doesn't, you can set `packageManager` in your `package.json`, remove the `yarnPath` setting (and file), and it'll work just fine as long as you have Corepack enabled. It's just that Yarn 4.0+ will do that automatically on `yarn init -2`.

**Comment by Rikaelus** [comment]
> It already doesn't, you can set `packageManager` in your `package.json`, remove the `yarnPath` setting (and file), and it'll work just fine as long as you have Corepack enabled. It's just that Yarn 4.0+ will do that automatically on `yarn init -2`.

This is a slight tangent and different situation, so if this is better it is own issue then just say so, but there's a lot of good context in this thread that makes for an excellent jumping-off point.

We're attempting to update a Docker-based project from having one of its services use Yarn 4 (upgrading from 3) and I'm really running afoul of the new Corepack requirements. Unlike @ildella we've never had to store the .cjs file in the repository but we do have tools in place that install it (and all other dependencies) when preparing the environment. That same tool also includes checks to make sure everything looks configured. The tool, for example, tests for the presence of `.yarn/releases/yarn-<ver>.cjs` and, if it's not present it automatically begins configuring it.

The first issue we had was this:
`HOME=/tmp/.yarn-global yarn set version 4.1.1` =>
```
error This project's package.json defines "packageManager": "yarn@4.1.1". However the current global version of Yarn is 1.22.22.

Presence of the "packageManager" field indicates that the project is meant to be used with Corepack, a tool included by default with all official Node.js distributions starting from 16.9 and 14.19.
Corepack must currently be enabled by running corepack enable in your terminal. For more information, check out https://yarnpkg.com/corepack. 
Failed
```

I got by that by changing the command to `HOME=/tmp/.yarn-global corepack yarn set version 4.1.1` but... I have no idea where Corepack is putting the .cjs file for our tool to see if it already exists. And I really need our project operating on files located only within the project directory structure for Docker voluming purposes. Can this be achieved? Is there any way to simply remove Corepack from the picture? From what I can tell it'll buy us nothing in a Dockerized environment but looks like it could cause us a good chunk of grief.

**Comment by koshic** [comment]
@Rikaelus,

1. Corepack and files location - see https://github.com/nodejs/corepack -> COREPACK_HOME. 
2. If you want to run your app inside pre-built container environment, you can 'install' (== put in the cache) more than one yarn version during container build: `RUN corepack enable yarn && corepack install --global --cache-only yarn@4.1.1 && <repeat it for any other version>`. As a result, '4.1.1' (and other versions you specified) will be available offline. If your app (== package.json) requires something else, it will be downloaded on the fly (use COREPACK_ENABLE_NETWORK to control network access).

**Comment by Rikaelus** [comment]
> 1. Corepack and files location - see https://github.com/nodejs/corepack -> COREPACK_HOME.
> 2. If you want to run your app inside pre-built container environment, you can 'install' (== put in the cache) more than one yarn version during container build: `RUN corepack enable yarn && corepack install --global --cache-only yarn@4.1.1 && <repeat it for any other version>`. As a result, '4.1.1' (and other versions you specified) will be available offline. If your app (== package.json) requires something else, it will be downloaded on the fly (use COREPACK_ENABLE_NETWORK to control network access).

Even with `corepack enable yarn && corepack install --global --cache-only yarn@4.1.1` and specifying `COREPACK_HOME` to be within the project root, it looks like Corepack is making functional changes outside our project root, for instance trying to unlink `/usr/local/bin/yarn`. That's the crux of my problem here. Previously everything necessary to use Yarn was contained within the project root that was carefully shared with the host machine and across containers as a Docker volume.

So it's not that I want to run the app inside a pre-built container environment. It's that we can use Yarn within a project root directory that's usable _across_ container environments without the wider environment changes that Corepack seems to be requiring. We really just don't need a package manager manager. We already take care of installing the Yarn version the project needs and in a way conducive to cross-container volumes.

**Comment by koshic** [comment]
> Even with `corepack enable yarn && corepack install --global --cache-only yarn@4.1.1` and specifying `COREPACK_HOME` to be within the project root, it looks like Corepack is making functional changes outside our project root, for instance trying to unlink `/usr/local/bin/yarn`

It not quite true :) Lets use the following container:
```dockerfile
FROM alpine:3.20.0

# git is required only for 'yarn init'
RUN apk add nodejs-current git && corepack enable
```

Then build & run it with RO filesystem expect /tmp and /root (we can use different user / home dir, sure):
```shell
docker build -t xxx . 
docker run -it --rm --read-only --tmpfs=/tmp --tmpfs=/root xxx sh -c '\
    mkdir /tmp/corepack; \
    export COREPACK_HOME=/tmp/corepack; \
    export COREPACK_ENABLE_DOWNLOAD_PROMPT=0; \
    mkdir /tmp/app; \
    cd /tmp/app; \
    export YARN_CACHE_FOLDER=./.yarn/cache; \
    yarn init -2; \
    ls -la; \
    cat package.json;'
```

No errors, no unexpected fs write:
```shell
➤ YN0000: · Yarn 4.2.2
➤ YN0000: ┌ Resolution step
➤ YN0000: └ Completed
➤ YN0000: ┌ Fetch step
➤ YN0000: └ Completed
➤ YN0000: ┌ Link step
➤ YN0000: └ Completed
➤ YN0000: · Done in 0s 29ms
total 440
drwxr-xr-x    4 root     root           220 Jun  4 19:57 .
drwxrwxrwt    5 root     root           100 Jun  4 19:57 ..
-rw-r--r--    1 root     root           134 Jun  4 19:57 .editorconfig
drwxr-xr-x    7 root     root           220 Jun  4 19:57 .git
-rw-r--r--    1 root     root           142 Jun  4 19:57 .gitattributes
-rw-r--r--    1 root     root           333 Jun  4 19:57 .gitignore
-rwxr-xr-x    1 root     root        425326 Jun  4 19:57 .pnp.cjs
drwxr-xr-x    3 root     root            80 Jun  4 19:57 .yarn
-rw-r--r--    1 root     root             6 Jun  4 19:57 README.md
-rw-r--r--    1 root     root            54 Jun  4 19:57 package.json
-rw-r--r--    1 root     root           290 Jun  4 19:57 yarn.lock
{
  "name": "app",
  "packageManager": "yarn@4.2.2"
}
```

To summarize: you have to add only one extra command to your base container - `corepack enable`.

**Comment by Rikaelus** [comment]
> To summarize: you have to add only one extra command to your base container - `corepack enable`.

It was quite a bit more complicated than that given the complexity of our environment but I finally got a working solution. The project root is used not only across independent Docker Compose configurations and containers but Yarn is interacted with through multiple Docker layers in our E2E and production build Dockerfiles. I basically had to enable Corepack and configure Yarn in every distinct environment that deals with the common project root so there's environmental parity, on top of refactoring our development tools to account for it.

Suffice it to say having everything self-contained within the project root was a much more portable solution with less configuration overhead. I can see the value of Corepack for certain development environments but I wholeheartedly believe that something like this shouldn't be forced on developers. There should be a way to opt-out at least.

**Comment by arcanis** [comment]
> There should be a way to opt-out at least.

There is - you can check-in your binary in your repository.

**Comment by Rikaelus** [comment]
> > There should be a way to opt-out at least.
> 
> There is - you can check-in your binary in your repository.

Which is a pretty lousy practice IMO, even if Yarn is relatively small.

All this being said, further testing to prepare this reply has actually revealed something else as the actual pivotal factor here; not Yarn 4 (or at least not only Yarn 4)

Here's `yarn set version 4.1.1` being run in different Node Docker images.

node:18-alpine:
```
reset-yarn-1  | ➤ YN0000: You don't seem to have Corepack enabled; we'll have to rely on yarnPath instead
reset-yarn-1  | ➤ YN0000: Downloading https://repo.yarnpkg.com/4.1.1/packages/yarnpkg-cli/bin/yarn.js
reset-yarn-1  | ➤ YN0000: Saving the new release in .yarn/releases/yarn-4.1.1.cjs
reset-yarn-1  | ➤ YN0000: Done with warnings in 0s 452ms
```

node:20-alpine:
```
error This project's package.json defines "packageManager": "yarn@4.1.1". However the current global version of Yarn is 1.22.22.
Presence of the "packageManager" field indicates that the project is meant to be used with Corepack, a tool included by default with all official Node.js distributions starting from 16.9 and 14.19.
Corepack must currently be enabled by running corepack enable in your terminal. For more information, check out https://yarnpkg.com/corepack.
```

node:21-alpine / node:22-alpine:
```
reset-yarn-1  | ➤ YN0000: You don't seem to have Corepack enabled; we'll have to rely on yarnPath instead
reset-yarn-1  | ➤ YN0000: Downloading https://repo.yarnpkg.com/4.1.1/packages/yarnpkg-cli/bin/yarn.js
reset-yarn-1  | ➤ YN0000: Saving the new release in .yarn/releases/yarn-4.1.1.cjs
reset-yarn-1  | ➤ YN0000: Done with warnings in 0s 232ms
reset-yarn-1 exited with code 0
```

We upgraded the images used to `node:20-alpine` along with our larger upgrade to Yarn 4, Next.js 14, etc. And it actually looks like it was `node:20-alpine` specifically that was forcing Corepack to be used. All the other versions before and after accepted Corepack not being enabled and installed Yarn within the project root as desired. I'm not sure if it was Node.js 20 or something else in the image, but whatever was forcing it seems to have been reverted.

Maybe this is something worth noting on https://yarnpkg.com/corepack that the `node:20-alpine` error points you to?

**Comment by abelbraaksma** [comment]
I'm stuck in a similar loop of damnation (but on Windows and Docker alike):

* With `yarnPath` set in `yarnrc.yml`, I will always receive `Error: Cannot find module 'D:\projects\.yarn\releases\yarn-4.6.0.cjs'`. Any `npm install` or `corepack enable && corepack install` has no effect.
* Without `yarnPath` set in `yarnrc.yml` and using `yarn set version 4.6.0` I get `error This project's package.json defines "packageManager": "yarn@4.6.0". However the current global version of Yarn is 1.22.22.` and this:

    > Presence of the "packageManager" field indicates that the project is meant to be used with Corepack, a tool included by default with all official Node.js distributions starting from 16.9 and 14.19.
    > Corepack must currently be enabled by running corepack enable in your terminal. For more information, check out https://yarnpkg.com/corepack. 
* Running `corepack enable` echoes nothing to the screen, has no effect on the above
* Running `corepack prepare yarn@4.6.0 --activate` returns `Preparing yarn@4.6.0 for immediate activation...`, but still nothing works. No error is echoed.
* Running `npm uninstall -g yarn` does not uninstall the global version of yarn mentioned above (i.e. version 1.22.22)
* Running `npm install -g corepack` echoes `changed 1 package`, but has otherwise no effect
* `npm -v` gives 11.0, `node -v` returns 22.13. 

Same effect on Docker using `node:22-alpine`, except that I have not gotten it working like @Rikaelus.

Only way to fix `yarn` is to manually copy `yarn-4.6.0.cjs` and set the `yarnPath`. 

I understand that `corepack` + `yarn` was supposed to work and make life simpler, but as it currently stands, it is NOT possible to use `corepack` on any system I own (at least it does not do as advertised, and when there are errors, it does not show the errors so I cannot fix anything).

I realize that I may not understand how it is supposed to work, but from the readmes and instructions I understand that it "should work" using `corepack enable && corepack prepare yarn@4.6.0 --activate`.

As others have commented above, I can either (1) add binaries to my Git repo (the horror!) or (2) rollback and use yarn 1.x (I really don't want to do either).

**Comment by clemyan** [comment]
Currently, there are only two officially supported ways to use Yarn Modern:

- Manage your global Yarn installation with Corepack (and *only* with Corepack)
- Include the Yarn binary in your project and run it via `yarnPath`. The Yarn binary is part of your project so you need to copy/move/deploy to anywhere you need to run yarn in the project.

There are other ways to use Yarn Modern but those are not officially supported and we can provide little to no support ("Here be dragons").

Given Corepack's current status we will probably distribute Yarn Modern in other ways in the future, but that probably won't happen until v5.

> * With `yarnPath` set in `yarnrc.yml`, I will always receive `Error: Cannot find module 'D:\projects\.yarn\releases\yarn-4.6.0.cjs'`

That means you didn't copy/preserve the Yarn binary with your project

> * `error This project's package.json defines "packageManager": "yarn@4.6.0". However the current global version of Yarn is 1.22.22`
> * Running `corepack enable` echoes nothing to the screen, has no effect on the above
> * Running `corepack prepare yarn@4.6.0 --activate` returns `Preparing yarn@4.6.0 for immediate activation...`, but still nothing works. No error is echoed.

That means your global Yarn installation is not managed by Corepack. I.e. you have a `yarn` binary in your `PATH` shadowing Corepack's `yarn` shim. Use `where yarn` on CMD or `(Get-Command yarn).Source` on PowerShell to find out which `yarn` binary your shell is using. If your global Yarn installation is managed by Corepack, it should be in the same directory as your `node` binary.

Regardless, if you have `yarnPath`, that will override Corepack. Once `yarnPath` is set, you need to bring the binary with you wherever it project goes until you remove `yarnPath`.

> * Running npm uninstall -g yarn does not uninstall the global version of yarn mentioned above (i.e. version 1.22.22)

That means your global Yarn installation is not managed by `npm` either. (If it were, `corepack enable` should overwrite it). So you need to find out which `yarn` binary your shell is using.
